base/src/main/java/io/spine/query/Subject.java

```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright 2020, TeamDev. All rights reserved.
+ *
+ * Redistribution and use in source and/or binary forms, with or without
+ * modification, must retain the above copyright notice and the following
+ * disclaimer.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.spine.query;
+
+import com.google.common.collect.ImmutableList;
+import com.google.protobuf.Message;
+
+import java.util.Objects;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+/**
+ * Set of criteria for the records obtained via querying or subscription.
+ *
+ * <p>Subjects are typically stored Protobuf messages or entity states.
+ *
+ * @param <I>
+ *         the type of the identifiers of the queried records
+ * @param <R>
+ *         the type of the queried records
+ */
+public final class Subject<I, R extends Message> {
+
+    /**
+     * The criteria put on the identifiers of the records of interest.
+     */
+    private final IdParameter<I> id;
+
+    /**
+     * The type of the queried records.
+     */
+    private final Class<R> recordType;
+
+    /**
+     * The type of the identifiers of the queried records.
+     */
+    private final Class<I> idType;
```

Comment:
Can we have `idType` declared before `id`? This way it would be easier to understand it as filtering thing.